Textmodel integration

Chandu-4444 opened the pull request:
Can do the following:

```julia
lm = FastText.LanguageModel(true)
classifier = FastText.TextClassifier(lm)
FastText.train_classifier!(classifier) # Would throw an error as I haven't fully enabled the model to work with FastAI's data container.
```

lorenzoh commented:
I think you've readded some files from TextModels.jl that we don't need, could you remove those? 🙂

Chandu-4444 commented:
Sure! Will clean up in the next commit.

Chandu-4444 commented:
- Deleted the files from TextModels.jl that are unnecessary.
- Add a utility function for padding the batches for sequence data (Very naively implemented, as a starting point though)

```julia
julia> batches = FastText.load_batchseq(data, task)
julia> batches[1][1]
92-element Vector{Vector{Int64}}:
 [25000, 25000, 25000, 25000, 25000, 25000, 25000, 25000]
 [633779, 633779, 633779, 633779, 633779, 633779, 633779, 633779]
 [2731, 34, 315, 354, 2087, 2209, 70, 1307]
 [44047, 435, 633779, 633779, 6589, 633779, 633779, 205]
 ⋮
 [0, 0, 0, 0, 0, 213, 0, 0]
 [0, 0, 0, 0, 0, 25, 0, 0]
 [0, 0, 0, 0, 0, 1778, 0, 0]

julia> batches[1][2]
8-element Vector{Int64}:
 1
 1
 1
 1
 1
 0
 1
 1
```

Chandu-4444 commented:
- fastai's way of encoding data doesn't include the removal of stop words (and Jeremy recommends it). So, I removed the stop word removal step.
- I've added a `vocab_size` keyword argument to `TextClassificationSingle`.
- Added `<unk>`, `<pad>` to the vocabulary.

Next:
- A batch loader for textdata with padding length = max(sentence length in a batch).

ToucheSir commented:
Should the vocab CSV files be checked in? I would've assumed they would be artifacts or DataDeps as well.

ToucheSir commented on FastText/src/models/train_text_classifier.jl:
```diff
@@ -0,0 +1,58 @@
+
+Flux.@functor TextClassifier
+
+function loss(m, xs, y; k = 10)
+    # forward steps
+    # Flux.reset!(m.rnn_layers)
+    Zygote.ignore() do
+        [m.rnn_layers(x) for x in xs[1:(end - k)]]
+    end
+    # bptt
+    ypreds = m.linear_layers([m.rnn_layers(x) for x in xs[(end - k + 1):end]])
+    loss = Flux.Losses.logitcrossentropy(ypreds, y)
+    println("Loss: $loss")
+    return loss
+end
+
+function train_text_classifier(classifier::TextClassifier = Nothing, batches = Nothing)
+    opt = Adam(1e-4)
+
+    for batch in batches
+        xs, y = batch
+        ps = Flux.params(classifier)
+        gs = gradient(() -> loss(classifier, batch...), ps)
+        Flux.Optimise.update!(opt, ps, gs)
+    end
+
+end
```
Are these temporary for testing the PR or do they come from TextModels? If the latter, I would get rid of them.

Chandu-4444 replied:
Yes, they're just temporary for training the model.

ToucheSir commented on FastText/src/models/pretrain_lm.jl:
```diff
@@ -0,0 +1,52 @@
+            VarDrop(layer_drop_prob),
+            AWD_LSTM(hid_lstm_sz, out_lstm_sz, hid_drop_prob; init = (dims...) -> init_weights(1/hid_lstm_sz, dims...)),
+            VarDrop(final_drop_prob),
+            x -> de(x, true),
```
```suggestion
            Base.Fix2(de, true),
```
Otherwise the model won't be serializable.

ToucheSir commented on FastText/src/models/pretrain_lm.jl:
```diff
@@ -0,0 +1,52 @@
+        Chain(
+            de,
+            VarDrop(in_drop_prob),
+            AWD_LSTM(embedding_size, hid_lstm_sz, hid_drop_prob; init = (dims...) -> init_weights(1/hid_lstm_sz, dims...)),
```
Likewise, I think this repeated lambda should be pulled out into a helper function of some sort.

ToucheSir commented on FastText/src/models/pretrain_lm.jl:
```diff
@@ -0,0 +1,52 @@
+end
+
+function LanguageModel(load_pretrained::Bool=false, task::Any = Nothing;embedding_size::Integer=400, hid_lstm_sz::Integer=1150, out_lstm_sz::Integer=embedding_size,
+    embed_drop_prob::Float64 = 0.05, in_drop_prob::Float64 = 0.4, hid_drop_prob::Float64 = 0.5, layer_drop_prob::Float64 = 0.3, final_drop_prob::Float64 = 0.3)
```
Enforcing all the drop probabilities to Float64 seems like a bad idea. We could either weaken it to something like `AbstractFloat` or drop the annotation altogether.

ToucheSir commented on FastText/src/models/pretrain_lm.jl:
```diff
@@ -0,0 +1,52 @@
+
+julia> lm = LanguageModel()
+"""
+mutable struct LanguageModel
```
```suggestion
struct LanguageModel
```
Both fields of this struct are abstract types, so we'll have to think about that sooner or later. Until then, can this be immutable?

ToucheSir commented on FastText/src/models/dataloader.jl:
```diff
@@ -0,0 +1,22 @@
+
+
+function load_sample(batches; batch_index=1, sample_seq_index=1, sample_index=1)
+    seq = []
+    for i in batches[batch_index][sample_seq_index]
+        seq = push!(seq, i[sample_index])
+    end
```
```suggestion
    seq = [i[sample_index] for i in batches[batch_index][sample_seq_index]]
```
This can be a comprehension.

ToucheSir commented on FastText/src/models/dataloader.jl:
```diff
@@ -0,0 +1,22 @@
+        seq = push!(seq, i[sample_index])
+    end
+
+    return (seq, batches[batch_index][2][sample_seq_index])
```
Some explanation of  what `[2]` means here would help. Giving it a variable name is the easiest choice, but a comment about the general expected input format of `batches` would be even better.

Chandu-4444 replied:
OK.

ToucheSir commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,410 @@
+
+This funciton works same as `params` function except for `AWD_LSTM` layer.
+While getting `Params` of the `AWD_LSTM` it does not include the `h` and `c` `params` of `AWD_LSTM`.
+This is useful while calculating gradients because calculating gradients for `h` and `c` fields
+in `AWD_LSTM` is unnecessary here.
+
+# Example:
+
+julia> layers = Chain(DroppedEmbeddings(4,5,0.2),
+                    AWD_LSTM(5, 3),
+                    Dense(3, 2),
+                    softmax
+                );
+julia> p1 = params(layers);
+julia> p2 = get_trainable_params(layers);
+
+julia> length(p1)
+8
+
+julia> length(p2)
+6
+
+`Params` from all the other layers are included in p2 except for `h` and `c`
+"""
+function get_trainable_params(layers)
+    p = []
+    function get_awd_params(awd::AWD_LSTM)
+        return [awd.layer.cell.Wi,
+        awd.layer.cell.Wh,
+        awd.layer.cell.b]
+    end
+    for layer in layers
+        layer isa Array || (layer = [layer])
+        for l in layer
+            l isa AWD_LSTM && (append!(p, get_awd_params(l)); continue)
+            push!(p, l)
+        end
+    end
+    return Flux.params(p...)
+end
```
I would get rid of this. If it's needed for the training code, we can discuss how to replace it.

Chandu-4444 replied:
That won't be necessary right now, I'll remove this.

ToucheSir commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,410 @@
+
+Flux.@functor DroppedEmbeddings
+
+Flux.trainable(m::DroppedEmbeddings) = (m.emb,)
```
```suggestion
Flux.trainable(m::DroppedEmbeddings) = (; emb=m.emb)
```

ToucheSir commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,410 @@
+"""
+asgd_step!(i, l) = nothing
+
+function asgd_step!(iter::Integer, layer::AWD_LSTM)
+    if (iter >= layer.T) & (layer.T > 0)
+        p = get_trainable_params([layer])
+        avg_fact = 1/max(iter - layer.T + 1, 1)
+        if avg_fact != 1
+            layer.accum = layer.accum .+ p
+            for (ps, accum) in zip(p, layer.accum)
+                ps .= avg_fact*accum
+            end
+        else
+            layer.accum = deepcopy(p)   # Accumulator for ASGD
+        end
+    end
+    return
+end
```
@lorenzoh I don't believe we have any utils for this in Flux or FastAI, right? Haven't kept up with the list of callbacks.

ToucheSir commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,410 @@
+
+AWD_LSTM(in::Integer, out::Integer, p::Float64=0.0; kw...) = AWD_LSTM(WeightDroppedLSTM(in, out, p; kw...), -1, [])
+
+Flux.@functor AWD_LSTM
+
+Flux.trainable(m::AWD_LSTM) = (m.layer,)
```
```suggestion
Flux.@functor AWD_LSTM (layer,)
```

ToucheSir reviewed (commented):
Porting the TextModels code is going to be a big task, but for now the priority is just to make things work e2e :)

Chandu-4444 commented:
```julia
julia> data, blocks = load(datarecipes()["imdb"])
((mapobs(loadfile, ObsView(::MLDatasets.FileDataset{typeof(identity), String}, ::Vector{Int64})), mapobs(parentname, ObsView(::MLDatasets.FileDataset{typeof(identity), String}, ::Vector{Int64}))), (Paragraph(), Label{String}(["neg", "pos"])))

julia> task = TextClassificationSingle(blocks, data)
SupervisedTask(Paragraph -> Label{String})

julia> model = FastAI.taskmodel(task, FastText.LanguageModel(false, task))
#90 (generic function with 1 method)

julia> batches = FastText.load_batchseq(data, task)
6250-element Vector{Tuple{Vector{Vector{Int64}}, WARNING: both Losses and NNlib export "ctc_loss"; uses of it in module Flux must be qualified
Flux.OneHotArray{UInt32, 2, 1, 2, Vector{UInt32}}}}:
 ([[35, 35, 35, 35], [3, 3, 3, 9], [40, 18025, 15, 14], [224, 10, 3541, 3040], [737, 34, 24, 505], [49, 7, 809, 3], [4, 4, 221, 3836], [1927, 104, 4, 
3], [7, 16, 629, 28440], [6, 351, 7, 17]  …  [2, 2, 2, 44], [2, 2, 2, 3], [2, 2, 2, 9839], [2, 2, 2, 17], [2, 2, 2, 1041], [2, 2, 2, 27], [2, 2, 2, 3], [2, 2, 2, 3836], [2, 2, 2, 3], [2, 2, 2, 28440]], [0 0 1 1; 1 1 0 0])

julia> using FluxTraining

julia> td, vd = splitobs(batches, at=0.9)

julia> using Flux

julia> learner = Learner(model, Flux.Losses.logitcrossentropy, callbacks=[Metrics(accuracy)]; data=(td, vd))
Learner()

julia> fit!(learner, 1)
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+    Wi::A
+    Wh::A
+    b::V
+    h::S
+    c::S
```
These are now replaced by `state0` and stored in `Recur`

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+reset_masks!(entity) = nothing
+reset_probability!(entity) = nothing
+
+"""
+    drop_mask(x, p)
+
+Drop mask generator
+
+This function generates dropout mask for given 'x' with `p` probability
+    or
+It can be used to generate the mask by giving the shape of the desired mask and probaility
+"""
+function drop_mask(x, p)
+    y = similar(x, size(x))
+    Flux.rand!(y)
+    y .= Flux._dropout_kernel.(y, p, 1 - p)
+    return y
+end
+
+drop_mask(shape::Tuple, p; type = Float32) = (mask = rand(type, shape...);mask .= _dropout_kernel.(mask, p, 1 - p))
```
Flux has `Flux.dropout_mask(x, p)` which replaces this

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+    maskWi::M
+    maskWh::M
+    active::Bool
+    state0
```
This needs a type parameter

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+end
+
+function WeightDroppedLSTMCell(in::Integer, out::Integer, p::Float64=0.0;
+    init = Flux.glorot_uniform)
```
We need `init`, `initb`, and `init_state` separately here

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+
+Flux.@functor WeightDroppedLSTMCell
+
+Flux.trainable(m::WeightDroppedLSTMCell) = (m.Wi, m.Wh, m.b, m.h, m.c)
```
```suggestion
Flux.trainable(m::WeightDroppedLSTMCell) = (m.Wi, m.Wh, m.b)
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+    return Flux.Recur(cell, hidden)
+end
+
+"""
+    reset!(m)
+
+Resets the h, c parameters of the LSTM Cell.
+
+For more refer [`Flux.reset`](@ref https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.reset!)
+"""
+function reset!(layers)
+    for layer in layers
+        if typeof(layer) == FastText.WeightDroppedLSTM
+            (layer.layer.state = (layer.layer.cell.h, layer.layer.cell.c))
+        else
+            Flux.reset!(layer)
+        end
+    end
+end
```
This can be deleted once the `state0` changes are incorporated correctly.

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+end
+
+
+"""
+    reset_masks!(layer)
+
+This is an important funciton since it used to reset the masks
+which are saved in WeightDroppedLSTMCell after every pass.
+
+julia> wd = WeightDroppedLSTM()
+
+julia> reset_masks!(wd)
+"""
+function reset_masks!(wd::T) where T <: Flux.Recur{<:WeightDroppedLSTMCell}
+    wd.cell.maskWi = drop_mask(wd.cell.Wi, wd.cell.p)
+    wd.cell.maskWh = drop_mask(wd.cell.Wh, wd.cell.p)
+    return
+end
```
This implies the masks should be part of the state?

darsnack replied:
```suggestion
function Flux.reset!(layer::Flux.Recur{<:WeightDroppedLSTMCell})
    maskWi = drop_mask(layer.cell.Wi, layer.cell.p)
    maskWh = drop_mask(layer.cell.Wh, layer.cell.p)
    layer.state = (layer.cell.state0..., maskWi, maskWh)
    
    return nothing
end
```

darsnack reviewed (changes requested):
Just reviewing the weight dropped LSTM for now

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+
+function (m::WeightDroppedLSTMCell)((h, c), x)
+    b, o = m.b, size(h, 1)
+    Wi = m.active ? m.Wi .* m.maskWi : m.Wi
+    Wh = m.active ? m.Wh .* m.maskWh : m.Wh
```
```suggestion
    Wi = Flux._isactive(m) ? m.Wi .* m.maskWi : m.Wi
    Wh = Flux._isactive(m) ? m.Wh .* m.maskWh : m.Wh
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+"""
+function WeightDroppedLSTM(a...; kw...)
+    cell = WeightDroppedLSTMCell(a...;kw...)
+    hidden = (cell.h, cell.c)
```
```suggestion
    maskWi = drop_mask(cell.Wi, cell.p)
    maskWh = drop_mask(cell.Wh, cell.p)
    hidden = (cell.state0..., maskWi, maskWh)
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+    h::S
+    c::S
+    p::Float64
+    maskWi::M
+    maskWh::M
```
Remove these fields too

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+        true,
+        (Flux.zeros32(out, 1), Flux.zeros32(out, 1))
+    )
+    cell.b[gate(out, 2)] .= 1
```
Why does it only mask part of the bias @Chandu-4444?

Chandu-4444 replied:
Since it is ported from TextModels.jl I'm not completely sure of this. But I'm sure this would be of the same as this https://github.com/FluxML/Flux.jl/blob/a377cb0637557b4c189fcc904bec65df9a124eb0/src/layers/recurrent.jl#L292.

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+        p,
+        drop_mask((out*4, in), p),
+        drop_mask((out*4, out), p),
+        true,
```
```suggestion
        :auto,
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+mutable struct DroppedEmbeddings{A, F}
+    emb::A
+    p::F
+    mask
```
This should have a type

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+            init(in, embed_size),
+            p,
+            drop_mask((in,), p),
+            true
```
```suggestion
            :auto
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+end
+
+function (de::DroppedEmbeddings)(x::AbstractArray, tying::Bool=false)
+    dropped = de.active ? de.emb .* de.mask : de.emb
```
```suggestion
    dropped = Flux._isactive(de) ? de.emb .* de.mask : de.emb
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+testmode!(m::DroppedEmbeddings, mode=true) =
+  (m.active = (isnothing(mode) || mode == :auto) ? nothing : !mode; m)
+
+function reset_masks!(de::DroppedEmbeddings)
```
Overload `Flux.reset!` here

darsnack replied:
Bumping this comment

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+
+julia> pd = PooledDense(40, 20)    # if the output size of the RNN layer is 40 in this case
+"""
+mutable struct PooledDense{F, S, T}
+    W::S
+    b::T
+    σ::F
+end
```
```suggestion
struct PooledDense{F, S, T}
    W::S
    b::T
    σ::F
end
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,296 @@
+PooledDense(W, b) = PooledDense(W, b, identity)
+
+function PooledDense(hidden_sz::Integer, out::Integer, σ = identity;
+             initW = Flux.glorot_uniform, initb = (dims...) -> zeros(Float32, dims...))
```
```suggestion
             initW = Flux.glorot_uniform, initb = Flux.zeros32)
```

ToucheSir commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,300 @@
+julia> reset_masks!(vd)
+"""
+
+mutable struct VarDrop{F}
+    p::F
+    mask
+    reset::Bool
+    active::Bool
+end
+
+VarDrop(p::Float64=0.0) = VarDrop(p, Array{Float32, 2}(UndefInitializer(), 0, 0), true, true)
+
+function (vd::VarDrop)(x)
+    vd.active || return x
+    if vd.reset
+        vd.mask = drop_mask(x, vd.p)
+        vd.reset = false
+    end
+    return (x .* vd.mask)
+end
```
As promised, here's the proposal from our call today:
```julia
mutable struct VarDrop{F}
    p::F
    active::Union{Bool, Nothing} # matches other norm layers
end

VarDrop(p::Real=0.0) = VarDrop(p, nothing)

function (vd::VarDrop)((has_mask, mask), x)
    if active
        mask = has_mask ? mask : drop_mask(x, vd.p)
        return (true, mask), x .* mask
    elseif !has_mask
        return (has_mask, mask), x
    else
        error("Mask set but layer is in test mode. Call `reset!` to clear the mask.")
    end
end
```
Plus bits to make `testmode!`, `_isactive` and `reset!` work properly.

Chandu-4444 commented:
The changes have been merged to #258.

darsnack commented on FastText/src/models/train_text_classifier.jl:
```diff
@@ -0,0 +1,60 @@
+
+To train create and instance and give it as first argument to 'train_classifier!' function
+"""
+mutable struct TextClassifier{A,F,Q}
```
```suggestion
struct TextClassifier{A,F,Q}
```

darsnack commented on FastText/src/models/train_text_classifier.jl:
```diff
@@ -0,0 +1,60 @@
+    )
+end
+
+Flux.@functor TextClassifier
```
```suggestion
Flux.@functor TextClassifier
Flux.trainable(m::TextClassifier) = (rnn_layers = m.rnn_layers, linear_layers = m.linear_layers)
```

Chandu-4444 replied:
No luck! The problem still remains the same.

darsnack reviewed (changes requested):
Try this

lorenzoh commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,267 @@
+mutable struct DroppedEmbeddings{A,F}
+    emb::A
+    p::F
+    mask::Vector{Float32}
```
This has to be an abstract array type so that GPU arrays (`CuVector`s) also work

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,267 @@
+"""
+function WeightDroppedLSTM(a...; kw...)
+    cell = WeightDroppedLSTMCell(a...; kw...)
+    maskWi = Flux.dropout_mask(Flux.rng_from_array(), cell.Wi, cell.p)
+    maskWh = Flux.dropout_mask(Flux.rng_from_array(), cell.Wh, cell.p)
```
```suggestion
    maskWi = Flux.dropout_mask(Flux.rng_from_array(cell.Wi), cell.Wi, cell.p)
    maskWh = Flux.dropout_mask(Flux.rng_from_array(cell.Wh), cell.Wh, cell.p)
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,267 @@
+end
+
+function Flux.reset!(layer::Flux.Recur{<:WeightDroppedLSTMCell})
+    maskWi = Flux.dropout_mask(Flux.rng_from_array(), layer.cell.Wi, layer.cell.p)
+    maskWh = Flux.dropout_mask(Flux.rng_from_array(), layer.cell.Wh, layer.cell.p)
```
```suggestion
    maskWi = Flux.dropout_mask(Flux.rng_from_array(layer.cell.Wi), layer.cell.Wi, layer.cell.p)
    maskWh = Flux.dropout_mask(Flux.rng_from_array(layer.cell.Wh), layer.cell.Wh, layer.cell.p)
```

darsnack commented on FastText/src/models/custom_layers.jl:
```diff
@@ -0,0 +1,267 @@
+
+function (vd::VarDropCell)((has_mask, mask), x)
+    if Flux._isactive(vd)
+        mask = has_mask ? mask : Flux.dropout_mask(Flux.rng_from_array(), x, vd.p)
```
```suggestion
        mask = has_mask ? mask : Flux.dropout_mask(Flux.rng_from_array(x), x, vd.p)
```

darsnack reviewed (changes requested):
Some of the issues that you posted on Zulip should be resolved by this
